platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java

```diff
@@ -117,21 +124,94 @@ public Collection<Result> readRequest(ExecutableElement method) {
-            String propertyName = getPropertyName(method);
-            Type originalType = extractOriginalType(method, annotation.get());
             AnnotationMirror annotationMirror = annotation.get();
-            boolean isFluentSetter = AnnotationUtils.findAnnotationValue(annotationMirror, "fluentSetter")
-                .map(v -> (Boolean) v.getValue())
-                .orElse(false);
-            CallInterceptionRequest groovyPropertyRequest = createGroovyPropertyInterceptionRequest(propertyName, method, originalType);
-            CallInterceptionRequest jvmGetterRequest = createJvmGetterInterceptionRequest(propertyName, method, originalType);
-            CallInterceptionRequest jvmSetterRequest = createJvmSetterInterceptionRequest(propertyName, method, originalType, isFluentSetter);
-            return Arrays.asList(new Success(groovyPropertyRequest), new Success(jvmGetterRequest), new Success(jvmSetterRequest));
+            List<AccessorSpec> accessorSpecs = readAccessorSpecsFromUpgradedProperty(method, annotationMirror);
+            List<CallInterceptionRequest> requests = new ArrayList<>();
+            for (AccessorSpec accessorSpec : accessorSpecs) {
+                switch (accessorSpec.accessorType) {
+                    case GETTER:
+                        if (isProviderOrProperty(accessorSpec.originalType)) {
+                            // We don't support Provider or Property as original type for a getter accessor
+                            return Collections.singletonList(new InvalidRequest(String.format("`Provider` or `Property` type is not supported as original type for a GETTER accessor. Method: %s.%s", method.getEnclosingElement(), method)));
+                        }
+                        CallInterceptionRequest groovyPropertyRequest = createGroovyPropertyInterceptionRequest(accessorSpec, method);
+                        CallInterceptionRequest jvmGetterRequest = createJvmGetterInterceptionRequest(accessorSpec, method);
+                        requests.add(groovyPropertyRequest);
+                        requests.add(jvmGetterRequest);
+                        break;
+                    case SETTER:
+                        CallInterceptionRequest jvmSetterRequest = createJvmSetterInterceptionRequest(accessorSpec, method);
+                        requests.add(jvmSetterRequest);
+                        break;
+                    default:
+                        throw new IllegalArgumentException("Unsupported accessor type: " + accessorSpec.accessorType);
+                }
+            }
+            return requests.stream()
+                .map(Success::new)
+                .collect(Collectors.toList());
         } catch (AnnotationReadFailure failure) {
             return Collections.singletonList(new InvalidRequest(failure.reason));
         }
     }
 
+    static boolean isProviderOrProperty(Type type) {
+        return type.equals(PROVIDER_TYPE) || type.equals(PROPERTY_TYPE);
+    }
```

Comment:
❓ We also have `RegularFileProperty` and `DirectoryProperty`, and potentially more in the future. Is there a better way we can handle this?

Reply:
I've removed support for Provider as original type. Currently it's not done well and kinda unsafe, since we don't have a get a parameter type of `Provider` if passed as `originalType = Provider.class`. If Java would support `originalType = Provider<File>.class` it would be much easier.

So I think we need to:
1. not support that, and do manual upgrades
2. support it differently, e.g. originalType is enum instead of class, so we can have `originalType = PROVIDER_OF_FILE`